# Описание проекта

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Описание данных

Таблица books
Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица authors
Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

Таблица publishers
Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства.

Таблица ratings
Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица reviews
Содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

# Цель проекта

Подготовить анализ базы данных для формирования ценностного предложения для нового продукта

# Задачи проекта

1. Посчитать, сколько книг вышло после 1 января 2000 года;
2. Посчитать для каждой книги количество обзоров и среднюю оценку;
3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
4. Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

# Исследование таблиц

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db']) 
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Таблица books

In [3]:
query = '''
SELECT 
    *
FROM books
'''

In [4]:
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

Таблица содержит данные о 1000 книг

## Таблица authors

In [5]:
query = '''
SELECT 
    *
FROM authors
'''

In [6]:
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

Таблица содержит данные о 636 авторах

## Таблица publishers

In [7]:
query = '''
SELECT 
    *
FROM publishers
'''

In [8]:
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

Таблица содержит данные о 340 издательствах

## Таблица ratings

In [9]:
query = '''
SELECT 
    *
FROM ratings
'''

In [10]:
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

Таблица содержит данные о 6456 пользовательских оценках книг

## Таблица reviews

In [11]:
query = '''
SELECT 
    *
FROM reviews
'''

In [12]:
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Таблица содержит данные о 2793 пользовательских обзорах книг

# Расчет, сколько книг вышло после 1 января 2000 года

In [13]:
query = '''
SELECT 
    COUNT(DISTINCT book_id) AS cnt
FROM 
    books
WHERE
    publication_date > '2000-01-01'
'''

In [14]:
pd.io.sql.read_sql(query, con = engine)

cnt
0  819

В базе данных содержится информация о 819 книгах вышедших после 01 января 2000 года

# Расчет для каждой книги количества обзоров и среднюю оценку

In [15]:
query = '''
SELECT
    books.title AS book_name,
    subquery_reviews.reviews_cnt AS reviews_cnt,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    LEFT JOIN (
        SELECT
            book_id,
            COUNT(review_id) AS reviews_cnt
        FROM
            reviews
        GROUP BY
            book_id
        ) AS subquery_reviews ON subquery_reviews.book_id = books.book_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
WHERE reviews_cnt IS NOT NULL
GROUP BY
    books.title,
    subquery_reviews.reviews_cnt
ORDER BY
    reviews_cnt DESC
'''

In [16]:
pd.io.sql.read_sql(query, con = engine)

book_name  reviews_cnt  \
0                              Twilight (Twilight  #1)            7   
1                                      Eat  Pray  Love            6   
2    Harry Potter and the Chamber of Secrets (Harry...            6   
3    Harry Potter and the Prisoner of Azkaban (Harr...            6   
4                            Outlander (Outlander  #1)            6   
..                                                 ...          ...   
989                                The Social Contract            1   
990  To Green Angel Tower (Memory  Sorrow  and Thor...            1   
991               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1            1   
992                              Wicked: The Grimmerie            1   
993                  Winter Prey (Lucas Davenport  #5)            1   

     avg_rating  
0      3.662500  
1      3.395833  
2      4.287500  
3      4.414634  
4      4.125000  
..          ...  
989    3.500000  
990    4.500000  
991    2.500000  
992    3.500000  
993    4.500000  

[994 rows x 3 columns]

In [17]:
query = '''
SELECT
    books.title AS book_name,
    subquery_reviews.reviews_cnt AS reviews_cnt,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    LEFT JOIN (
        SELECT
            book_id,
            COUNT(review_id) AS reviews_cnt
        FROM
            reviews
        GROUP BY
            book_id
        ) AS subquery_reviews ON subquery_reviews.book_id = books.book_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
    books.title,
    subquery_reviews.reviews_cnt
ORDER BY
    avg_rating DESC
'''

In [18]:
pd.io.sql.read_sql(query, con = engine)

book_name  reviews_cnt  \
0                                  Homage to Catalonia          2.0   
1                                                March          2.0   
2    In the Hand of the Goddess (Song of the Liones...          2.0   
3                 Welcome to Temptation (Dempseys  #1)          2.0   
4    Stone of Farewell (Memory  Sorrow  and Thorn  #2)          1.0   
..                                                 ...          ...   
995  The World Is Flat: A Brief History of the Twen...          3.0   
996                                              Junky          2.0   
997                  His Excellency: George Washington          2.0   
998                                      Drowning Ruth          3.0   
999                               Harvesting the Heart          2.0   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

На основании анализа данных можно сделать вывод: количество обзоров книг не очень большое, максимальное количество обзоров на одну книгу составляет семь, а шесть книг вообще не имеют обзоров

# Определить издательство, которое выпустило наибольшее число книг толще 50 страниц

In [19]:
query = '''
SELECT
    publishers.publisher,
    COUNT(book_id) AS books_cnt
FROM
    books
    LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE 
    num_pages > 50
GROUP BY
    publishers.publisher
ORDER BY
    books_cnt DESC
LIMIT 1
'''

In [20]:
pd.io.sql.read_sql(query, con = engine)

publisher  books_cnt
0  Penguin Books         42

На основании анализа данных видим, что в базе содержатся больше всего книг выпущенных издательством "Penguin Books" - 42 книги. Это британское издательсво довольно давно рабаотает на рынке - с 1935 года

# Определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [21]:
query = '''
SELECT
    authors.author,
    AVG(subquery_ratings.avg_rating) AS avg_rating
FROM
    books
    LEFT JOIN (
        SELECT 
            book_id,
            AVG(rating) AS avg_rating,
            COUNT(rating_id) AS rating_cnt
        FROM
            ratings
        GROUP BY
            book_id
    ) AS subquery_ratings ON subquery_ratings.book_id = books.book_id
    LEFT JOIN authors ON authors.author_id = books.author_id
WHERE
    rating_cnt >= 50
GROUP BY
    author
ORDER BY
    avg_rating DESC
LIMIT 1
'''

In [22]:
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

На основании анализа можно сделать вывод, самый высокий рейтинг книг имеет Джоан Роулинг. Гниги о Гарри Поттере достаточно популярны во всем мире

# Расчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок.

In [23]:
query = '''
WITH
    ratings_subquery AS (
        SELECT
            ratings.username,
            COUNT(rating_id) AS ratings_cnt
        FROM
            ratings
        GROUP BY
            ratings.username
        HAVING
            COUNT(rating_id) > 50
        ),
    reviews_subquery AS (
        SELECT
            reviews.username,
            COUNT(review_id) AS reviews_cnt
        FROM
            reviews
        GROUP BY
            reviews.username
    )
SELECT
    AVG(reviews_cnt) as avg_reviews_cnt
FROM
    ratings_subquery
    LEFT JOIN reviews_subquery ON reviews_subquery.username = ratings_subquery.username
'''

In [24]:
pd.io.sql.read_sql(query, con = engine)

avg_reviews_cnt
0        24.333333

На основании анализа данных видим, что наиболее активные пользователи сервиса в среднем оставляют 24 обзора на книги

# Общие выводы

В результате проекта проведен анализ базы данных содержащей информацию о сервисе чтения книг по подписке. База данных содержит информацию о книгах, авторах, издательствах, оценках пользователей и обзоры на книги. На основании проведенного анализа можно сделать следующие выводы:
1. База данных сервиса содержит 1000 книг;
2. Книги 636 авторов и выпущенные 340 издательсвами;
3. В базе 80% книг изданы после 01 января 2000 года
4. Максимальное количесво обзоров оставлено на книгу "Twilight"
5. Среди книг толще 50 сраниц, больше всего выпущены издательсвом "Penguin Books"
6. Сымый высокий рейтинг у автора Джоан Роулинг. Средняя оценка пользователей - 4,28
7. Пользователи сервиса достаточно активные, достаточно часто оставляю обзоры на прочитанные книги.